### PR TITLE
Minimal API for COLR/CPAL

### PR DESCRIPTION
**Pull request by khaledhosny**
WIP, the tests do not pass (table in cpal-v1.ttf fails sanitization). Here is a LibreOffice using it (WIP as well):
image

### CODE REVIEW

File: src/hb-ot-color.cc [hunk 1]
```diff
@@ -146,38 +158,58 @@ hb_ot_color_get_palette_flags (hb_face_t *face, unsigned int palette)
...
-				unsigned int    *color_count /* IN/OUT */,
-				hb_ot_color_t   *colors /* OUT */)
+hb_ot_color_get_palette_colors (hb_face_t      *face,
+				unsigned int    palette, /* default=0 */
+				unsigned int    start_offset,
+				unsigned int   *count /* IN/OUT */,
+				hb_ot_color_t  *colors /* OUT */)
 {
   const OT::CPAL& cpal = _get_cpal(face);
-  if (unlikely (palette >= cpal.numPalettes))
```
Comment (ebraminio) [inline src/hb-ot-color.cc]: This work is great, thanks!

I believe you can move the logic inside this and `hb_ot_color_get_color_layers` to the table itself and use this as a proxy only.

**Comment by khaledhosny** [comment]
Closing, not sure about this any more. Can be done from scratch if we ever want it.

**Comment by behdad** [comment]
Why?!!  I definitely want something like this.  Leave open, it's fine.

**Comment by khaledhosny** [comment]
I thought we had another PR from @ebraminio for that, but looks like we don’t.

**Comment by ebraminio** [comment]
Khaled and Behdad any chance you can see if is this good enough?

My question, are hb_ot_color_has_{cpal,colr}_data needed? May `hb_ot_color_t` be renamed to `hb_color_t`? Is `hb_ot_color_get_{red,green,blue,alpha}` correct name for public macros on harfbuzz?

**Comment by khaledhosny** [comment]
The macros should be all caps I think.

**Comment by khaledhosny** [comment]
`has_*_data`would be used by clients to see if the font can decompose color glyphs or not, and it can check that once instead of calling the color API on each glyph and checking the result.

**Comment by ebraminio** [comment]
Makes enough sense, thanks

**Comment by behdad** [comment]
Don't merge any API in please until we get the 2.0.0 post-releases out first.

**Comment by ebraminio** [comment]
> Don't merge any API in please until we get the 2.0.0 post-releases out first.

Sure.

How about have it as a branch on the repo itself?

Asking from Khaled, any personal reason marking it as WIP? Please rename the macros if you like to

File: src/dump-emoji.cc [hunk 2]
```diff
@@ -120,50 +127,62 @@ static void colr_cpal_rendering (cairo_font_face_t *cairo_face, unsigned int upe
...
+	  for (unsigned int layer = 0; layer < num_layers; ++layer)
+	  {
+	    uint32_t color = 0xFF;
+            if (color_indices[layer] != 0xFFFF)
+	      color = colors[color_indices[layer]];
+	    int alpha = hb_ot_color_get_alpha (color);
+	    int r = hb_ot_color_get_red (color);
+	    int g = hb_ot_color_get_green (color);
+	    int b = hb_ot_color_get_blue (color);
+	    cairo_set_source_rgba (cr, r / 255., g / 255., b / 255., alpha);
```
Comment (ebraminio) [inline src/dump-emoji.cc]: My fault probably, this needs `/ 255.` also I guess

**Comment by khaledhosny** [comment]
@ebraminio feel free to close this PR, move the branch to harfbuzz/harfbuzz repo and open a new one. I don’t seem to remember much of what I was doing here, so feel free to fix as you see fit.

**Comment by ebraminio** [comment]
Merged on ot-color (not master, don't get confused)

**Pull request merged**